### PR TITLE
Convert to systemd resources

### LINKED ISSUE #628: Request to remove runit
### Cookbook version
5.0.2

### Chef-client version
Chef Development Kit Version: 1.4.3
chef-client version: 12.19.36
delivery version: master (41b94ffb5efd33723cf72a89bf4d273c8151c9dc)
berks version: 5.6.4
kitchen version: 1.16.0
inspec version: 1.25.1

### Platform Details
Windows 10, converging a Centos 7.3.1611 node

### Scenario:
Converge works the first time, but subsequent converges fail when trying to install plugins because the Jenkins service isn't started.  The only solution is to `kitchen destroy` and `kitchen converge` after making any change.

### Steps to Reproduce:
Using the Jenkins WAR file and a list of plugins.  `kitchen converge` the first time will be successful, immediately running `kitchen converge` a second, third, forth time will fail because the Jenkins service isn't running.

### Expected Result:
Jenkins to behave like it did after the first `kitchen converge`.

### Actual Result:
Error during `kitchen converge` because the Jenkins service isn't running.

Runit has not been updated in almost 3 years.  We were able to confirm that the issue I am having is not a direct issue with this cookbook, but a bug in runit.  When I do a `kitchen converge` for the first time successfully, then do a `kitchen login`, if I then manually stop the service and then try to manually restart it, runit still says that Jenkins is down.  This is effectively the same thing I noticed when I do a converge subsequent times.  The cookbook rightly tells runit to start again, but runit fails to start the service.  I am running a clean install of Centos 7.3.1611 when I try this.  I would like to see this cookbook remove the runit dependency and rely on the default init system.  Other cookbooks have gone this route lately, like [the version 6.0.0 of the nginx_chef cookbook](https://github.com/chef-cookbooks/chef_nginx/blob/578cfeb25237a672ea6bbe261fca89281195a27c/CHANGELOG.md) released this spring.

### PR DESCRIPTION
**Pull request by detjensrobert**
Signed-off-by: Robert Detjens <detjensrobert@osuosl.org>

# Description

Converts the runit services to proper systemd resources and removes dependency on the runit cookbook.

While this cookbook is due for a table flip, this PR does not do that.

### Deprecations

- `jenkins_slave_jnlp`:
  - property `runit_groups` renamed to `service_groups`

- runit needs manual cleanup / uninstallation

## Issues Resolved

Closes #628

## Check List

- [x] All tests pass. See TESTING.md for details.
- [ ] New functionality includes testing.
- [ ] New functionality has been documented in the README if applicable.

### Tesing note

Sometimes the first `jenkins_plugin` install fails, likely do to marketplace data not being ready or similar. A second converge passes, though.

### CODE REVIEW

**Comment by kitchen-porter** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This is a big Pull Request.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: CHANGELOG.md [hunk 1]
```diff
@@ -4,6 +4,14 @@ This file is used to list changes made in each version of the jenkins cookbook.
 
 ## Unreleased
 
+- Remove runit dependency
+- Use systemd units instead of runit services
```
Comment (majormoses) [inline CHANGELOG.md:8]: As we are not handling the conversion on a running system I consider this a breaking change hence the major release label change. If we were to stop the existing `runit` service prior to bringing up the new service and provide a clean migration path we could release this as a minor release.

**Review by majormoses — APPROVED** [review]
LGTM

IMO the changelog needs a bit of enhancement. For example it does not highlight breaking changes such as dropping support of older chef versions.

**Comment by nuclearsandwich** [comment]
@detjensrobert I'm really excited by this PR! I still need to do a deeper read through but I do have a few broad questions:

* The README was reformatted with a hard wrap, was that in response to a specific style guide or just an editor default? It would be easier to review the content changes if the line wrapping was either not changed or was changed separately.

* The heredoc syntax changes can probably also be moved into a separate PR for ease of review and to make sure that the commit messages match the changes.

* This is more of a general proposition to the other sous-chefs but does it make sense to use this major version change to update the file names in this cookbook for current [Jenkins terminology](https://github.com/jenkinsci/jep/blob/master/jep/16/README.adoc) such as agent and controller?

**Comment by detjensrobert** [comment]
- The README was an editor setting -- I'll revert that
- I'll move those heredocs to another PR as well :+1:

File: CHANGELOG.md (hunk 1 above)
Comment (ramereth) [reply CHANGELOG.md:8]: i.e. I think you should move this into the section below so folks know it's a breaking change.

**Review by ramereth — COMMENTED** [review]
I almost feel as though we may need an `UPGRADING.md` doc for folks migrating from runit to systemd. But I'm also not sure if we need to do that.  I feel like we should have _something_ for those users on how to migrate.

**Comment by detjensrobert** [comment]
Looking at the runit cookbook, the service resource only has `:stop` and `:disable`. If we keep the runit dependency for another release, we can stop/disable the old service and start the new one. The runit files will need to be manually cleaned up still since the runit cookbook doesn't do that. 

There does not seem to be a clean way to uninstall runit, since the runit cookbook uses a recipe to install and it up.

Should this PR try to do this or leave the cleanup for the admin since its not a complete job?

**Comment by ramereth** [comment]
@detjensrobert IMO we should at least leave some basic instructions for how to migrate but we shouldn't need to have a working solution within the cookbook for the migration.

**Comment by Xorima** [comment]
Normal sous-chefs convention is to have an `UPGRADING.md` file to cover the how to do this (normally it's a copy of the text from the changelog, but sometimes more)

**Comment by davidsainty** [comment]
I'm also really interested to see this in, but in my environment we have a substantial number of projects with their own Jenkins servers, all converging and continuously integrating off the same Jenkins cookbook.  With a manual upgrade path, I'm really not sure that makes upgrading this cookbook even possible - coordinating all projects at one time is just not going to happen.

I haven't tested the outcome of trying to converge this over a running system, but it looks like it might be messy.

Perhaps just accepting that converging over a runit-based installation is too hard and just leaving it alone - until the next full rebuild, or until the upgrade process is followed - is better?

**Comment by detjensrobert** [comment]
I'm currently doing some tests with converging over an existing node. I'll get an `UPGRADING.md` file going once I figure out what all is needed.

**Comment by nuclearsandwich** [comment]
Does it improve the manual upgrade story to keep the runit-based recipe in for an additional major release but default to the systemd-based recipe so that projects can be on the same version but switch to systemd based service management when it suits them?

@davidsainty would that help your situation at all or are the environments identically configured as well as sharing cookbook version?

**Comment by davidsainty** [comment]
Our systems are all very differently configured.

I think the key for us is probably that a converge on a runit-based build needs to be supported by default.  Support in this case probably only means that it doesn't break it, which in turn probably means it should either leave the existing runit-based service alone (by default), or directly support it (by default).

Honestly, I think I'm happy with the runit-based installation support being removed, and just detecting and treating a converge on a runit-based server as a special case where systemd configuration is not_if'd away...

**Comment by davidsainty** [comment]
> Does it improve the manual upgrade story to keep the runit-based recipe in for an additional major release but default to the systemd-based recipe so that projects can be on the same version but switch to systemd based service management when it suits them?

... and to answer this point directly, I think the complication would be that if the default is always "systemd" then there's still a substantial effort that would need to be coordinated just getting people to request the "runit" support be enabled over the default.  That effort alone would outweigh the value of upgrading the cookbook in the first place.

If the default was "auto" (auto-select runit if in use, or systemd if nothing is installed yet) then everything is good :)

**Comment by detjensrobert** [comment]
Did some testing getting a runit->systemd migration done, which seems doable (assuming the actual init/service manager is systemd and not fully runit). The runit-managed services are under `/etc/service` and `/etc/init.d`. Removing those files allows the new systemd service to start. 

The runit services are managed by a `runit` / `runsvdir-start` systemd service. Should this also stop/disable/mask this runit manger service, or should be left as a manual cleanup step?

**Comment by kitchen-porter** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This is a big Pull Request.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

**Comment by davidsainty** [comment]
I think in theory that other services might be installed via runit, though in practice it's probably very unlikely, so so long as Jenkins is shut down and not started any more via runit, any other runit-related leftovers are probably ok.

The important thing is that only one Jenkins is left running at the end, and as the converge against Jenkins is normally also executed from Jenkins that started via runit, there's something of a risk of the system dying on the self-surgery table.

I'll look to test this tomorrow.

**Comment by davidsainty** [comment]
Ok, so initial testing left me with a dead Jenkins service :)

This first problem is that this is typical:

force_default['jenkins']['master']['jvm_options'] = "-Djava.io.tmpdir='/var/tmp/directory/location' ... and some other flags"

Amazingly, the ExecStart in the jenkins.service does not handle that shell-like, and those quotes end up in java.io.tmpdir.  This causes horribly confusing error messages, but strace made it all clear.

If I remove the single quotes in the jenkins.service file the service can be started.

**Comment by davidsainty** [comment]
My goodness, I have a new-found loathing of systemd now :)  Its quote parsing is not very /bin/sh like.

I did come up with this though, which I think will reduce the number of failures with this change.

```diff
--- ../jenkins-detjensrobert-systemd/recipes/_master_war.rb     2021-07-15 08:16:21.000000000 +1200
+++ jenkins/recipes/_master_war.rb      2021-07-15 22:47:44.375923700 +1200
@@ -85,6 +85,12 @@
 #   action [:stop, :disable]
 # end
 
+#
+# systemd ExecStart does not fully support shell-style quoting as
+# might be expected by jvm_options etc.  But it does support a leading
+# quote, trailing quote, and escaping in between, and we can ask
+# /bin/sh to make sense of the rest.
+#
 systemd_unit 'jenkins.service' do
   content <<~EOU
     #
@@ -103,7 +109,7 @@
     Environment="JENKINS_HOME=#{node['jenkins']['master']['home']}"
     WorkingDirectory=#{node['jenkins']['master']['home']}
     #{ulimits_to_systemd(node['jenkins']['master']['ulimits'])}
-    ExecStart=#{node['jenkins']['java']} #{node['jenkins']['master']['jvm_options']} -jar jenkins.war --httpPort=#{node['jenkins']['master']['port']} --httpListenAddress=#{node['jenkins']['master']['listen_address']} #{node['jenkins']['master']['jenkins_args']}
+    ExecStart=/bin/sh -c 'exec #{"#{node['jenkins']['java']} #{node['jenkins']['master']['jvm_options']} -jar jenkins.war --httpPort=#{node['jenkins']['master']['port']} --httpListenAddress=#{node['jenkins']['master']['listen_address']} #{node['jenkins']['master']['jenkins_args']}".gsub("'","\\\\'")}'
 
     [Install]
     WantedBy=multi-user.target
```

**Comment by davidsainty** [comment]
I have seen at least one failed Jenkins restart due to the listening socket not completely freeing up in time.  But that experience is not too bad, re-running chef-client at least can repair that.  The patch solves the problem of having to try and debug why Jenkins won't start at all - and that debugging is a pretty nasty experience - mainly because the Jenkins error reports are lacking in precision - so I think that issue is worth fixing.

**Comment by ramereth** [comment]
@davidsainty I think this PR is ready to go. Can you please take another look at it? (The current failures are a race condition and unrelated to this pr).

**Comment by davidsainty** [comment]
> @davidsainty I think this PR is ready to go. Can you please take another look at it? (The current failures are a race condition and unrelated to this pr).

Yes I agree.  I haven't done a fresh test, but I did a test with the code as it stood + my patch, which I believe is exactly how the current PR stands, so I have no objections to merging.

**Review by ramereth — APPROVED** [review]

**Comment by kitchen-porter** [comment]
Released as: [9.0.0](https://github.com/sous-chefs/jenkins/releases/tag/9.0.0)

**Comment by nuclearsandwich** [comment]
Congrats @davidsainty! Good ship

**Comment by davidsainty** [comment]
@detjensrobert deserves the credit :)

**Pull request merged**